tensorly/tenalg/proximal.py

```diff
@@ -987,37 +987,23 @@ def hals_nnls(
         rec_error = 0
         for k in range(rank):
             if UtU[k, k]:
-                if (
-                    sparsity_coefficient is not None
-                ):  # Modifying the function for sparsification
-                    deltaV = tl.where(
-                        (UtM[k, :] - tl.dot(UtU[k, :], V) - sparsity_coefficient)
-                        / UtU[k, k]
-                        > -V[k, :],
-                        (UtM[k, :] - tl.dot(UtU[k, :], V) - sparsity_coefficient)
-                        / UtU[k, k],
-                        -V[k, :],
-                    )
-                    V = tl.index_update(V, tl.index[k, :], V[k, :] + deltaV)
+                term = UtM[k, :] - tl.dot(UtU[k, :], V)
 
-                else:  # without sparsity
-                    deltaV = tl.where(
-                        (UtM[k, :] - tl.dot(UtU[k, :], V)) / UtU[k, k] > -V[k, :],
-                        (UtM[k, :] - tl.dot(UtU[k, :], V)) / UtU[k, k],
-                        -V[k, :],
-                    )
-                    V = tl.index_update(V, tl.index[k, :], V[k, :] + deltaV)
+                # Modifying the function for sparsification
+                if sparsity_coefficient is not None:
```

Comment:
This code should be unchanged, but (1) reuses some indexing results for indexing-slow backend performance, and (2) I find it a bit more readable by removing repetitive segments and breaking the expressions up into logical units.